### ISSUE TITLE
Networking broken on master node after deploying hostNetwork=true Pod.

### ISSUE THREAD

**Issue by kupson** [k3s-io/k3s#3535]
**Environmental Info:**
K3s Version:
```
k3s version v1.21.2+k3s1 (5a67e8dc)
go version go1.16.4
```

Node(s) CPU architecture, OS, and Version: 
```
Linux kpi1 5.10.17-v8+ #1414 SMP PREEMPT Fri Apr 30 13:23:25 BST 2021 aarch64 GNU/Linux
```

Cluster Configuration:
- 1 server (aarch64)
- 4 agents (3 aarch64 + 1 amd64)

**Describe the bug:**
Broken network - normally I can ping 8.8.8.8 just fine.

```sh
PING 8.8.8.8 (8.8.8.8) 56(84) bytes of data.
From 192.168.0.131 icmp_seq=1 Destination Port Unreachable
ping: sendmsg: Operation not permitted
```

**Steps To Reproduce:**

- Installed K3s: k3s binary downloaded via `curl`
- Systemd file for server (**disabled `traefik` and `servicelb`**):
    ```
    [Unit]
    Description=Lightweight Kubernetes
    Documentation=https://k3s.io
    After=network-online.target
    Wants=network-online.target

    [Service]
    Type=notify
    EnvironmentFile=/etc/rancher/k3s/k3s_env
    PassEnvironment=K3S_TOKEN
    ExecStart=/usr/local/bin/k3s server --disable traefik,servicelb
    KillMode=process
    Delegate=yes
    LimitNOFILE=infinity
    LimitNPROC=infinity
    LimitCORE=infinity
    TasksMax=infinity
    TimeoutStartSec=0
    Restart=always
    RestartSec=5s

    [Install]
    WantedBy=multi-user.target
    ```
- Systemd file for agent differs in `ExecStart`
    ```
    ExecStart=/usr/local/bin/k3s agent --with-node-id --server https://${K3S_PRI_MASTER}:6443
    ```
- Deployed Pod **with `hostNetwork=true` and `nodeSelector` pointing to server node** via `kubectl apply -f -` (stdin below):
   ```
   {
      "apiVersion": "v1",
      "kind": "Pod",
      "metadata": {
         "creationTimestamp": null,
         "labels": {
            "run": "busybox"
         },
         "name": "busybox"
      },
      "spec": {
         "containers": [
            {
               "args": [
                  "sleep",
                  "600"
               ],
               "image": "busybox",
               "name": "busybox",
               "resources": { }
            }
         ],
         "dnsPolicy": "ClusterFirst",
         "hostNetwork": true,
         "nodeSelector": {
            "kubernetes.io/ingress": "true"
         },
         "restartPolicy": "Never"
      },
      "status": { }
   }
   ```
- Pod didn't start - couldn't pull image:
   ```
   NAME      READY   STATUS             RESTARTS   AGE
   busybox   0/1     ImagePullBackOff   0          12m
   ```

**Expected behavior:**

Pod running fine.

**Actual behavior:**
Broken network on server node.

**Additional context / logs:**
Network fixed after `kubectl delete pod busybox`.

**Comment by kupson** [comment]
Removing all NetworkPolicies also fixes the issue, it's the combination of NetPols and hostNetwork=true that results in broken network. (Not sure if it's only broken when deploying to server node).

NetPols:
```
- apiVersion: networking.k8s.io/v1
  kind: NetworkPolicy
  metadata:
    annotations: {}
    creationTimestamp: "2021-06-29T14:51:14Z"
    generation: 1
    labels:
      name: default-allow-dns
    name: default-allow-dns
    namespace: traefik-ingress
  spec:
    egress:
    - ports:
      - port: 53
        protocol: UDP
      - port: 53
        protocol: TCP
      to:
      - namespaceSelector:
          matchLabels:
            name: kube-system
        podSelector:
          matchExpressions:
          - key: k8s-app
            operator: In
            values:
            - kube-dns
    podSelector: {}
    policyTypes:
    - Egress
- apiVersion: networking.k8s.io/v1
  kind: NetworkPolicy
  metadata:
    annotations: {}
    creationTimestamp: "2021-06-29T14:51:15Z"
    generation: 1
    labels:
      name: default-deny-all
    name: default-deny-all
    namespace: traefik-ingress
  spec:
    podSelector: {}
    policyTypes:
    - Ingress
    - Egress
```

**Comment by brandond** [comment]
Are you also running Docker, NetworkManager, ufw/firewalld, or anything else that is attempting to manage iptables rules?

**Comment by kupson** [comment]
Nope, it's Raspberry Pi OS Lite (based on Debian Buster) with networking configured via `dhcpcd` on eth0. The `wpa_supplicant` is running but wlan0 is down. Legacy version of iptables is used.

[iptables-save-broken.txt](https://github.com/k3s-io/k3s/files/6746786/iptables-save-broken.txt)
[iptables-save-working.txt](https://github.com/k3s-io/k3s/files/6746787/iptables-save-working.txt)

There are no differences in *nat* and *mangle* tables. In *filter* it looks like the `k3s` is trying to apply NetPols to the `hostNetwork=true` Pod with disastrous effects.

The `192.168.0.131` is server node IP address.

`KUBE-ROUTER-*` chains:
```
-A KUBE-ROUTER-FORWARD -s 192.168.0.131/32 -m physdev --physdev-is-bridged -m comment --comment "rule to jump traffic from POD name:busybox namespace: traefik-ingress to chain KUBE-POD-FW-VT4J2LHIJ337GGVJ" -j KUBE-POD-FW-VT4J2LHIJ337GGVJ
-A KUBE-ROUTER-FORWARD -d 192.168.0.131/32 -m physdev --physdev-is-bridged -m comment --comment "rule to jump traffic destined to POD name:busybox namespace: traefik-ingress to chain KUBE-POD-FW-VT4J2LHIJ337GGVJ" -j KUBE-POD-FW-VT4J2LHIJ337GGVJ
-A KUBE-ROUTER-FORWARD -d 192.168.0.131/32 -m comment --comment "rule to jump traffic destined to POD name:busybox namespace: traefik-ingress to chain KUBE-POD-FW-VT4J2LHIJ337GGVJ" -j KUBE-POD-FW-VT4J2LHIJ337GGVJ
-A KUBE-ROUTER-FORWARD -s 192.168.0.131/32 -m comment --comment "rule to jump traffic from POD name:busybox namespace: traefik-ingress to chain KUBE-POD-FW-VT4J2LHIJ337GGVJ" -j KUBE-POD-FW-VT4J2LHIJ337GGVJ
-A KUBE-ROUTER-INPUT -s 192.168.0.131/32 -m comment --comment "rule to jump traffic from POD name:busybox namespace: traefik-ingress to chain KUBE-POD-FW-VT4J2LHIJ337GGVJ" -j KUBE-POD-FW-VT4J2LHIJ337GGVJ
-A KUBE-ROUTER-OUTPUT -d 192.168.0.131/32 -m comment --comment "rule to jump traffic destined to POD name:busybox namespace: traefik-ingress to chain KUBE-POD-FW-VT4J2LHIJ337GGVJ" -j KUBE-POD-FW-VT4J2LHIJ337GGVJ
-A KUBE-ROUTER-OUTPUT -s 192.168.0.131/32 -m comment --comment "rule to jump traffic from POD name:busybox namespace: traefik-ingress to chain KUBE-POD-FW-VT4J2LHIJ337GGVJ" -j KUBE-POD-FW-VT4J2LHIJ337GGVJ
``` 

The `KUBE-POD-FW-VT4J2LHIJ337GGVJ` chain:
```
-A KUBE-POD-FW-VT4J2LHIJ337GGVJ -m comment --comment "rule for stateful firewall for pod" -m conntrack --ctstate RELATED,ESTABLISHED -j ACCEPT
-A KUBE-POD-FW-VT4J2LHIJ337GGVJ -d 192.168.0.131/32 -m comment --comment "rule to permit the traffic traffic to pods when source is the pod\'s local node" -m addrtype --src-type LOCAL -j ACCEPT
-A KUBE-POD-FW-VT4J2LHIJ337GGVJ -m comment --comment "run through nw policy default-allow-dns" -j KUBE-NWPLCY-KLPEIZULGQEAZIGB
-A KUBE-POD-FW-VT4J2LHIJ337GGVJ -m comment --comment "run through nw policy default-deny-all" -j KUBE-NWPLCY-3TBMK5XQ7YVMOGOD
-A KUBE-POD-FW-VT4J2LHIJ337GGVJ -m comment --comment "rule to log dropped traffic POD name:busybox namespace: traefik-ingress" -m mark ! --mark 0x10000/0x10000 -m limit --limit 10/min --limit-burst 10 -j NFLOG --nflog-group 100
-A KUBE-POD-FW-VT4J2LHIJ337GGVJ -m comment --comment "rule to REJECT traffic destined for POD name:busybox namespace: traefik-ingress" -m mark ! --mark 0x10000/0x10000 -j REJECT --reject-with icmp-port-unreachable
-A KUBE-POD-FW-VT4J2LHIJ337GGVJ -j MARK --set-xmark 0x0/0x10000
```

**Comment by brandond** [comment]
Hmm, that is interesting. We use the kube-router network policy controller, you might check with them to see if this has been fixed upstream. You could also try using a commit build from PR https://github.com/k3s-io/k3s/pull/3557

**Comment by kupson** [comment]
Should be fixed upstream - https://github.com/cloudnativelabs/kube-router/issues/1067

**Comment by rancher-max** [comment]
I'm going to close this as it has been fixed and confirmed as part of #3373
